### ISSUE TITLE
Support RS485 control of Huanyang VFD (chinese spindles)

### ISSUE THREAD

**Issue by Harvie** [terjeio/grblHAL#68]
Hello,
can you please add support for this?

http://fightpc.blogspot.com/2014/10/vfd-control-with-arduino-using-rs485.html

**Comment by terjeio** [comment]
I can make a plugin. You can test as I do not own one? If so which driver to start with?

**Comment by Harvie** [comment]
Yes, i can test it :-)

I guess the C driver from LinuxCNC should be safe bet:
https://github.com/LinuxCNC/linuxcnc/blob/master/src/hal/user_comps/huanyang-vfd/hy_vfd.c

Or maybe you can try to merge this code from GRBL_ESP32:
https://github.com/davehines91/Grbl_Esp32/blob/Grbl_ESP32_RS485_VFD/Grbl_Esp32/VFD.cpp

But there are drivers in other languages as well:
https://github.com/itschleemilch/huanyango (GO)
https://github.com/GilchristT/SpindleTalker2 (C#)
https://github.com/drogenlied/pyHYControl (Python)


There are also guys who tried to get this merged to original GRBL, but there was not enough space on atmega:
https://github.com/bdring/Grbl_Esp32/issues/24
https://github.com/grbl/grbl/issues/516

**Comment by bdring** [comment]
https://github.com/bdring/Grbl_Esp32/blob/master/Grbl_Esp32/Spindles/HuanyangSpindle.cpp

**Comment by terjeio** [comment]
> Yes, i can test it :-)

For which processor/board then? I am not going to add this to all drivers even if it is going to be a plugin.

A [NucleoF411RE](https://www.st.com/en/evaluation-tools/nucleo-f411re.html) board and a Teensy 4.1 is in the mail so it could be one of those, but it will be at least a week before I have any chance of looking into this.

**Comment by Harvie** [comment]
> For which processor/board then? I am not going to add this to all drivers even if it is going to be a plugin.

I have "esp32 devkit v1"

![image](https://user-images.githubusercontent.com/276504/85687829-81c7da80-b6d1-11ea-8f69-cfc5ac92b42d.png)

**Comment by phil-barrett** [comment]
Is the interface just standard serial to an RS485 converter? Are there any other control signals needed?  My T4.x boards have a serial out connector.

**Comment by terjeio** [comment]
> Are there any other control signals needed?

Not if the converter has [auto direction](https://www.allaboutcircuits.com/industry-articles/simplify-your-industrial-interface-by-eliminating-rs-485-control-lines1/) circuitry.

>  My T4.x boards have a serial out connector.

All drivers has at least one serial port available, those that has serial over USB (or network connectivity) as well may assign the serial port to a secondary function such as a MPG or ModBus. Since RS485 VFD plugins should not handle the serial port peripheral directly, I am considering adding a driver level HAL to let the driver allocate a port and provide a standardized interface.

The main challenge could be how to handle the silence required around ModBus messages in a generic way, preferably in the plugin code...

**Comment by Harvie** [comment]
I have only MAX485 module available for testing, it has following pins at uC side: DI, DE, RE, RO

![image](https://user-images.githubusercontent.com/276504/85825622-fc046780-b782-11ea-84b0-b9d5105edd7f.png)

**Comment by terjeio** [comment]
@Harvie Which pin mappings are you using for your board?

I have started coding and are ready to start simulating but hit a snag as the messaging protocol is not ModBus, it is a chinese lookalike. This means my plan to code a ModBus emulation for the VFD is not possible, I have to write an emulation from scratch instead. I have a ModBus stack available so a real ModBus VFD emulator would be easy to write, but no luck with that...

Also, error handling needs to be implemented - it seems to me that implementation for the @bdring ESP32 port is rather crude in that respect, it would silently continue on a timeout or on an exception returned from the VFD (assuming the chinese ModBus lookalike might return exceptions per the ModBus standard). Executing g-code that makes chips with a non-running spindle is IMO not a good idea...

**Comment by Harvie** [comment]
I think for tests we can use mapping of this board:

https://www.tindie.com/products/33366583/grbl_esp32-cnc-development-board-v41/
https://github.com/bdring/Grbl_Esp32/wiki/Setting-Up-the-I-O-Pins#spindle

It does use following pinout:

```
#define SPINDLE_PWM_PIN         GPIO_NUM_25
#define SPINDLE_ENABLE_PIN      GPIO_NUM_22
#define SPINDLE_DIR_PIN         GPIO_NUM_18
```

So i guess these pins can be reused for digital communication instead... Is the pinout user-configurable? (at least during compilation) if i later decide to change it?

**Comment by Harvie** [comment]
> I have started coding and are ready to start simulating but hit a snag as the messaging protocol is not ModBus, it is a chinese lookalike.

Yes, it uses RS485 physical layer, but the protocol might be rather chinese... Earlier i've sent links to code which implements the protocol. I am not sure how different from ModBus it can be, maybe there's some compatible subset, which might be enough for basic communicatiobn. But that's just guess.

> Also, error handling needs to be implemented

What happens if you ignore this? I think there should be no collisions on the line as long as there is only VFD and ESP32, which is probably typical setup. I expect VFD to not send any status reports unless polled, but i am not sure... If such collisions are occuring, i think it might be enough to send every command 3 times with small random delay no matter if collision occured.

**Comment by terjeio** [comment]
> Is the pinout user-configurable?

Yes, there are board map files that may be changed or new ones created. I have to revisit the ESP32 code to see what limits apply. Note I am doing the initial coding for a processor that has a debugger.

> I am not sure how different from ModBus it can be

It deviates enough to prompt me to make an emulator from scratch, I have the basics working now - I only need to add fault scenarioes to make it complete?

The plugin also supports [P2x Huanyang VFD's](https://github.com/bdring/Grbl_Esp32/issues/453) which seems to use a 100% ModBus compliant protocol.

Error handling:

Status reports has to be requested so no risk of collisions with a correctly configured setup. What I was thinking of is what to do if the VFD does not respond as expected or not at all. IMO that is a critical error.

A complicating factor is that communication with the VFD can happen in an ISR context, and at least the dynamic RPM change in [stepper.c](https://github.com/terjeio/grblHAL/blob/master/grbl/stepper.c) is time critical. grblHAL has support for CSS (Constant Surface Speed) mode that uses this. Waiting for a response from the VFD in time critical code is a big no-no. Then there is the ESP32 that simply crashes if a float variable (RPM) is referenced in an ISR context, I see that Bart has changed rpm to an int in his port - I assume to avoid this.

Another is that the spindle will be stopped on a reset (mc_reset) which also might happen in an ISR context, e.g. when a hard limit is triggered.

All in all this is a harder task to get 100% right than I first thought...

**Comment by Harvie** [comment]
> The plugin also supports P2x Huanyang VFD's which seems to use a 100% ModBus compliant protocol.

This manual seems to apply https://wiki.spoje.net/lib/exe/fetch.php/howto/mechanical_engineering/huanyang_vfd_inverter_manual.pdf
On page 55 of that manual there are details about MODBUS implementation if that helps.

Hardware i have available for testing looks like this, i am not sure which model exactly it is. But it is 1.5kW HuanYang.

![image](https://user-images.githubusercontent.com/276504/87029807-f0805a00-c1e0-11ea-833e-32c607ab23a1.png)

**Comment by Harvie** [comment]
Also according to manual it seems that following stuff can be set:

```
Communication Address 1-250

Baud Rate 4800 b/s 9600 b/s 19200 b/s 34800 b/s 

Communication Data Method 
0：8N1 For ASCII 1：8E1 For ASCII 2：8O1 For ASCII
3：8N1 For RTU 4：8E1 For RTU 5：8O1 For RTU 

The communication protocol has two kinds of control mode：
⑴ RTU (Remote Terminal Unit) mode
⑵ ASCII（American Standard Code for information interchange）mode
Information of codes:
RTU mode: Each of 8-bit data is composed of two 4-bit (hexadecimal), for example: 64H
ASCII mode: Each of 8-bit data is composed of two ASCⅡbyte, for example:
One 1-bit data 64H (hexadecimal) i
```

**Comment by terjeio** [comment]
I have reorganized the core grblHAL a bit code in order to avoid setting spindle parameters from an ISR context and have successfully tested functionality with my VFD emulator (on a TM4C123 board) and development board (MSP432).

So am I now ready to add the plugin to the ESP32 driver. However the board you suggested does not have spindle direction pin available:

> I think for tests we can use mapping of this board:
https://www.tindie.com/products/33366583/grbl_esp32-cnc-development-board-v41/

If I am not mistaken this uses the [3axis_v4](https://github.com/bdring/Grbl_Esp32/blob/master/Grbl_Esp32/Machines/3axis_v4.h) pin mapping file.

An addtional pin is required for VFD comms, which one to sacrifice? The Mist pin?

Note that I have not yet verified that the pins available are possible to use for serial comms.

**Comment by Harvie** [comment]
Yes, i think the mist might be sacrificed for now as i do not have misting device and i definitely do not plan to use it during the spindle communication test.

**Comment by terjeio** [comment]
The [plugin code](https://github.com/terjeio/grblHAL/tree/test/plugins/spindle) is now ready for testing in the [development branch](https://github.com/terjeio/grblHAL/tree/test) - only for the ESP32 driver for now.

Settings can be found here:

https://github.com/terjeio/grblHAL/blob/9845e693296553caff13fcaf5c6a6aff3743152d/drivers/ESP32/bdring_v4_map.h#L98-L103

I have only tested with boosterpack_map.h which has different pin allocations.

The plugin supports P2x spindles as well, but I have not yet made that an option in the CMakeLists.txt.

Note that I am not yet 100% happy with the way ESP32 interacts with the plugin but with luck the core functionality should work. Among the issues I have with this driver is low polling frequency and the snowflake guru crashing the app if a float is encountered in an ISR context. To be resolved later.

**Comment by Harvie** [comment]
Sorry i was bit busy and totaly forgot to test this... Should be able to setup the test in few days...

**Comment by Harvie** [comment]
I am kinda confused about how to build grblHAL for ESP32... There does not seem to be `examples` directory in esp32 driver in order to be able to compile it using arduino IDE as described here https://github.com/terjeio/grblHAL/wiki/Compiling-GrblHAL

I might use make or cmake until arduino IDE support is added for ESP32, but i am not sure how...

**Comment by terjeio** [comment]
It has to be built with ESP-IDF v3.2 or later - but not v4+. I do not use Arduino libraries so not sure it can be built using the Arduino IDE, or what it would take to convert it for that. The project is set up for cmake.

https://github.com/terjeio/grblHAL/tree/master/drivers/ESP32

**Comment by Harvie** [comment]
I guess i still miss something...

```
[harvie@anemophobia ESP32]$ pwd
/home/harvie/Temp/grblHAL/drivers/ESP32
[harvie@anemophobia ESP32]$ cmake .
CMake Error at CMakeLists.txt:88 (include):
  include could not find load file:

    /tools/cmake/project.cmake


-- The C compiler identification is GNU 10.2.0
-- The CXX compiler identification is GNU 10.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:92 (target_compile_definitions):
  Cannot specify compile definitions for target "grbl.elf" which is not built
  by this project.


CMake Error at CMakeLists.txt:93 (target_compile_definitions):
  Cannot specify compile definitions for target "grbl.elf" which is not built
  by this project.


CMake Error at CMakeLists.txt:136 (target_add_binary_data):
  Unknown CMake command "target_add_binary_data".


-- Configuring incomplete, errors occurred!

```

**Comment by Harvie** [comment]
> I do not use Arduino libraries so not sure it can be built using the Arduino IDE, or what it would take to convert it for that.

I beleive it's merely about providing the directory structure and library.properties file expected by arduino IDE, since the ESP32 support package for arduino is based on ESP-IDF and there's not really a reason to use the Arduino compatible API wrapper layer, because ESP-IDF API can be called directly from Arduino IDE as well...

**Comment by terjeio** [comment]
Are you compiling with ESP-IDF?

> I beleive it's merely about providing the directory structure and library.properties file expected by arduino IDE

You can do that then?

**Comment by Harvie** [comment]
> Are you compiling with ESP-IDF?

Where do i put the ESP-IDF source code?

**Comment by Harvie** [comment]
Ok, i've managed to setup esp-idf, hopefully it will build now...

**Comment by Harvie** [comment]
Seems  to compile, but it does not link:

```
...
Scanning dependencies of target __ldgen_output_esp32.project.ld
[ 99%] Generating ld/esp32.project.ld
[ 99%] Built target __ldgen_output_esp32.project.ld
[ 99%] Generating ap_login.html.S
[ 99%] Generating favicon.ico.S
[ 99%] Generating index.html.S
Scanning dependencies of target grbl.elf
[ 99%] Building C object CMakeFiles/grbl.elf.dir/project_elf_src_esp32.c.obj
[ 99%] Building ASM object CMakeFiles/grbl.elf.dir/favicon.ico.S.obj
[ 99%] Building ASM object CMakeFiles/grbl.elf.dir/index.html.S.obj
[ 99%] Building ASM object CMakeFiles/grbl.elf.dir/ap_login.html.S.obj
[ 99%] Linking CXX executable grbl.elf
/home/harvie/.espressif/tools/xtensa-esp32-elf/esp-2020r3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/freertos/libfreertos.a(port_common.c.obj):(.literal.main_task+0x14): undefined reference to `app_main'
/home/harvie/.espressif/tools/xtensa-esp32-elf/esp-2020r3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/freertos/libfreertos.a(port_common.c.obj): in function `main_task':
/home/harvie/esp/esp-idf/components/freertos/port/port_common.c:129: undefined reference to `app_main'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/grbl.elf.dir/build.make:465: grbl.elf] Chyba 1
make[1]: *** [CMakeFiles/Makefile2:2246: CMakeFiles/grbl.elf.dir/all] Chyba 2
make: *** [Makefile:149: all] Chyba 2
```

**Comment by Harvie** [comment]
With ESP-IDF master (4.x) it builds but does not link, with v3.2 i did not tried to build at all, since they use different build system and i would have need to setup things differently.

With ESP-IDF v3.3 i get this:

```
[harvie@anemophobia ESP32]$ pwd
/home/harvie/Temp/grblHAL/drivers/ESP32
[harvie@anemophobia ESP32]$ idf.py build
Note: You are using Python 3.8.6. Python 3 support is new, please report any problems you encounter. Search for 'Setting the Python Interpreter' in the ESP-IDF docs if you want to use Python 2.7.
Checking Python dependencies...
Python requirements from /home/harvie/esp/esp-idf/requirements.txt are satisfied.
Running cmake in directory /home/harvie/Temp/grblHAL/drivers/ESP32/build
Executing "cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DESP_PLATFORM=1 --warn-uninitialized /home/harvie/Temp/grblHAL/drivers/ESP32"...
Warn about uninitialized values.
-- Found Git: /usr/bin/git (found version "2.29.2") 
-- IDF_TARGET not set, using default target: esp32
-- Building for target esp32
-- ccache will be used for faster builds
-- The C compiler identification is GNU 5.2.0
-- The CXX compiler identification is GNU 5.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /home/harvie/.espressif/tools/xtensa-esp32-elf/1.22.0-97-gc752ad5-5.2.0/xtensa-esp32-elf/bin/xtensa-esp32-elf-gcc
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/harvie/.espressif/tools/xtensa-esp32-elf/1.22.0-97-gc752ad5-5.2.0/xtensa-esp32-elf/bin/xtensa-esp32-elf-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /home/harvie/.espressif/tools/xtensa-esp32-elf/1.22.0-97-gc752ad5-5.2.0/xtensa-esp32-elf/bin/xtensa-esp32-elf-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building empty aws_iot component due to configuration
-- Component names: soc log heap freertos vfs newlib esp_ringbuf driver esp_event ethernet mbedtls micro-ecc efuse bootloader_support partition_table app_update spi_flash nvs_flash lwip tcpip_adapter pthread smartconfig_ack wpa_supplicant xtensa-debug-module espcoredump esp32 cxx app_trace asio jsmn aws_iot bootloader nimble bt coap console nghttp esp-tls esp_adc_cal tcp_transport esp_http_client esp_http_server esp_https_ota openssl esp_https_server esp_websocket_client esptool_py expat wear_levelling sdmmc fatfs freemodbus idf_test json libsodium mdns mqtt protobuf-c protocomm spiffs ulp unity wifi_provisioning
-- Component paths: /home/harvie/esp/esp-idf/components/soc;/home/harvie/esp/esp-idf/components/log;/home/harvie/esp/esp-idf/components/heap;/home/harvie/esp/esp-idf/components/freertos;/home/harvie/esp/esp-idf/components/vfs;/home/harvie/esp/esp-idf/components/newlib;/home/harvie/esp/esp-idf/components/esp_ringbuf;/home/harvie/esp/esp-idf/components/driver;/home/harvie/esp/esp-idf/components/esp_event;/home/harvie/esp/esp-idf/components/ethernet;/home/harvie/esp/esp-idf/components/mbedtls;/home/harvie/esp/esp-idf/components/micro-ecc;/home/harvie/esp/esp-idf/components/efuse;/home/harvie/esp/esp-idf/components/bootloader_support;/home/harvie/esp/esp-idf/components/partition_table;/home/harvie/esp/esp-idf/components/app_update;/home/harvie/esp/esp-idf/components/spi_flash;/home/harvie/esp/esp-idf/components/nvs_flash;/home/harvie/esp/esp-idf/components/lwip;/home/harvie/esp/esp-idf/components/tcpip_adapter;/home/harvie/esp/esp-idf/components/pthread;/home/harvie/esp/esp-idf/components/smartconfig_ack;/home/harvie/esp/esp-idf/components/wpa_supplicant;/home/harvie/esp/esp-idf/components/xtensa-debug-module;/home/harvie/esp/esp-idf/components/espcoredump;/home/harvie/esp/esp-idf/components/esp32;/home/harvie/esp/esp-idf/components/cxx;/home/harvie/esp/esp-idf/components/app_trace;/home/harvie/esp/esp-idf/components/asio;/home/harvie/esp/esp-idf/components/jsmn;/home/harvie/esp/esp-idf/components/aws_iot;/home/harvie/esp/esp-idf/components/bootloader;/home/harvie/esp/esp-idf/components/nimble;/home/harvie/esp/esp-idf/components/bt;/home/harvie/esp/esp-idf/components/coap;/home/harvie/esp/esp-idf/components/console;/home/harvie/esp/esp-idf/components/nghttp;/home/harvie/esp/esp-idf/components/esp-tls;/home/harvie/esp/esp-idf/components/esp_adc_cal;/home/harvie/esp/esp-idf/components/tcp_transport;/home/harvie/esp/esp-idf/components/esp_http_client;/home/harvie/esp/esp-idf/components/esp_http_server;/home/harvie/esp/esp-idf/components/esp_https_ota;/home/harvie/esp/esp-idf/components/openssl;/home/harvie/esp/esp-idf/components/esp_https_server;/home/harvie/esp/esp-idf/components/esp_websocket_client;/home/harvie/esp/esp-idf/components/esptool_py;/home/harvie/esp/esp-idf/components/expat;/home/harvie/esp/esp-idf/components/wear_levelling;/home/harvie/esp/esp-idf/components/sdmmc;/home/harvie/esp/esp-idf/components/fatfs;/home/harvie/esp/esp-idf/components/freemodbus;/home/harvie/esp/esp-idf/components/idf_test;/home/harvie/esp/esp-idf/components/json;/home/harvie/esp/esp-idf/components/libsodium;/home/harvie/esp/esp-idf/components/mdns;/home/harvie/esp/esp-idf/components/mqtt;/home/harvie/esp/esp-idf/components/protobuf-c;/home/harvie/esp/esp-idf/components/protocomm;/home/harvie/esp/esp-idf/components/spiffs;/home/harvie/esp/esp-idf/components/ulp;/home/harvie/esp/esp-idf/components/unity;/home/harvie/esp/esp-idf/components/wifi_provisioning
-- IDF_VER: v3.3.4-262-ga19f58355
-- Project version: 377742b
-- Found PythonInterp: /home/harvie/.espressif/python_env/idf3.3_py3.8_env/bin/python (found version "3.8.6") 
-- Found Perl: /usr/bin/perl (found version "5.32.0") 
-- Adding linker script /home/harvie/Temp/grblHAL/drivers/ESP32/build/esp-idf/esp32/esp32_out.ld
-- Adding linker script /home/harvie/esp/esp-idf/components/esp32/ld/esp32.rom.ld
-- Adding linker script /home/harvie/esp/esp-idf/components/esp32/ld/esp32.peripherals.ld
-- Adding linker script /home/harvie/esp/esp-idf/components/esp32/ld/esp32.rom.libgcc.ld
-- Adding linker script /home/harvie/esp/esp-idf/components/esp32/ld/esp32.rom.spiram_incompatible_fns.ld
-- Building empty aws_iot component due to configuration
-- Component libraries: 
-- Configuring done
CMake Error at /home/harvie/esp/esp-idf/tools/cmake/project.cmake:103 (add_executable):
  Cannot find source file:

    eeprom/eeprom_24LC16B.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
  .hpp .hxx .in .txx
Call Stack (most recent call first):
  CMakeLists.txt:90 (project)


CMake Error at /home/harvie/esp/esp-idf/tools/cmake/project.cmake:103 (add_executable):
  No SOURCES given to target: grbl.elf
Call Stack (most recent call first):
  CMakeLists.txt:90 (project)


CMake Generate step failed.  Build files cannot be regenerated correctly.
cmake failed with exit code 1
```

**Comment by terjeio** [comment]
You have to start the "ESP-IDF Command Prompt", `cd` to the project folder and build with `idf.py build`. Or you can build and flash with `idf.py -p <port> flash` 

I have tried several times to compile from Eclipse which is my preferred development platform, but have given up. PlatformIO next? I need a new development machine for that - no disk space for that in my aging machine...

Jus saw your new comment:

missing file: eeprom/eeprom_24LC16B.c

Remove these two lines from CMakeLists.txt:

 eeprom/eeprom_24LC16B.c
 eeprom/eeprom_24AAxxx.c

Perhaps I should port to v4, the current build system is IMO trash (or I am stupid in not understanding it).

**Comment by Harvie** [comment]
That is what i do...

i've done `. export.sh` (which setups the commandline environment) then `cd` to project directory and then `idf.py build`  (or `mkdir build; cd build; cmake ..` which seems to do the same thing)

**Comment by terjeio** [comment]
FYI I just built and flashed my ESP32 (with v3.3).

Any helpful info here?

https://docs.espressif.com/projects/esp-idf/en/v3.2.2/get-started/index.html#get-started-get-esp-idf

**Comment by terjeio** [comment]
Ouch, I did no see your latest comment... Sorry for the noise.

**Comment by Harvie** [comment]
BTW i've changed the baudrate to 9600, which is probably better for high EMI environment that is usualy near the VFD stuff and rebuilt the firmware. Is it ok, that other defines did not changed?

```
# grep -ri baud .  | grep -i modb
./driver.c:#ifndef MODBUS_BAUD
./driver.c:#define MODBUS_BAUD 19200
./driver.c:    serial2Init(MODBUS_BAUD);
./bdring_v4_map.h:#define MODBUS_BAUD             9600
./cnc_boosterpack_map.h:#define MODBUS_BAUD             19200
```

**Comment by Harvie** [comment]
Anyway it's great that i was able to build this. I can now debug bCNC to fix grblHAL compatibility issues that people report...

And of course i will test the RS485 spindle compatibility soon...

**Comment by Harvie** [comment]
So i've compiled the code, uploaded firmware to esp32, hooked logic analyzer to pins 2, 21 and 22 and i do not see any digital signal when i issue `M3 S12000` or `M5` g-code command...

**Comment by Harvie** [comment]
Any chances i do need to configure anything else than just copying in these files?

```
$ ls spindle/
huanyang.c  huanyang.h  modbus.c  modbus.h  README.md
```

**Comment by terjeio** [comment]
> So i've compiled the code, uploaded firmware to esp32, hooked logic analyzer to pins 2, 21 and 22 and i do not see any digital signal when i issue M3 S12000 or M5 g-code command...

Which board map file? Seems it is only _bdring_v4_map.h_ that has the driver configs for ModBus. But looks like you have figured this out already.

_my_machine.h_ contains settings not covered by _CMakeLists.txt_.

> Any chances i do need to configure anything else than just copying in these files?

You must enable Huanyang spindle support in _CMakeLists.txt_.

---

I am just back home, it has been a busy day today so will revisit/recheck this tomorrow.

**Comment by Harvie** [comment]
> You must enable Huanyang spindle support in CMakeLists.txt.

Switched from OFF to ON:

```
../esp32-hal-uart.c: In function 'uartConfig':
../esp32-hal-uart.c:213:33: error: 'UART2_TX_PIN' undeclared (first use in this function)
         uart_set_pin(uart->num, UART2_TX_PIN, UART2_RX_PIN, UART_PIN_NO_CHANGE, UART_PIN_NO_CHANGE);
                                 ^
../esp32-hal-uart.c:213:33: note: each undeclared identifier is reported only once for each function it appears in
../esp32-hal-uart.c:213:47: error: 'UART2_RX_PIN' undeclared (first use in this function)
         uart_set_pin(uart->num, UART2_TX_PIN, UART2_RX_PIN, UART_PIN_NO_CHANGE, UART_PIN_NO_CHANGE);
                                               ^

```

**Comment by Harvie** [comment]
i've added following to `esp32-hal-uart.c`

```C
#define UART2_RX_PIN            GPIO_NUM_22
#define UART2_TX_PIN            GPIO_NUM_21
#define MODBUS_DIRECTION_PIN    GPIO_NUM_2
#define MODBUS_BAUD             9600
```
But i still get an error:

```
../driver.c: In function 'driver_setup':
../driver.c:1347:62: error: 'SPINDLE_MASK' undeclared (first use in this function)
         .pin_bit_mask = DIRECTION_MASK|STEPPERS_DISABLE_MASK|SPINDLE_MASK|COOLANT_MASK,
                                                              ^
../driver.c:1347:62: note: each undeclared identifier is reported only once for each function it appears in
../driver.c: In function 'vModBusPollCallback':
../driver.c:1456:5: error: implicit declaration of function 'modbus_poll' [-Werror=implicit-function-declaration]
     modbus_poll();
     ^

```

**Comment by terjeio** [comment]
I'll take a look at this tomorrow, be patient. It is likely I have messed up a bit since the grblHAL core has had a lot of changes lately to facilitate better plugin support. So sorry for stressing you about this without verifying the code first.

**Comment by Harvie** [comment]
Don't worry. I am very grateful for your effort and some testing is the least i can do. Also it is my fault that i've not tested the code back then when it was compatible.

**Comment by terjeio** [comment]
[Test branch](https://github.com/terjeio/grblHAL/tree/test) updated with fix for the regression (it was mainly due to core changes). With a little luck it should now work, at least my simulator accepts what I believe are the correct commands.

Note that I have moved more options to CMakeLists.txt for simpler configuration when cmake is used for compilation.

**Comment by Harvie** [comment]
> Test branch updated with fix for the regression 

Ok, this builds, but when i enable HUANYANG, it fails to communicate with sender over serial line. theres nothing shown in serial console...

**Comment by terjeio** [comment]
Can you post the `$I` output when it works please. And did you update the plugin code?

**Comment by Harvie** [comment]
>  And did you update the plugin code?

stupid me :-) i'll try now

**Comment by Harvie** [comment]
Nice! Now i have some serial data on the output pins! But when i click spindle enable in bCNC the GRBL reports Alarm:14, perhaps because spindle is not connected. Hopefully i will try tomorow with real spindle to see if it works...

![image](https://user-images.githubusercontent.com/276504/99919613-cdcd8880-2d1e-11eb-8c15-5231d5ec0111.png)

**Comment by Harvie** [comment]
So just tested with spindle. It works like 10% of the time, but i get lot of alarms. error:9 and error:14 mostly...

**Comment by Harvie** [comment]
It seems to me like some RPM settings do work, while others does not. weird.
Sometimes alarm is caused, sometimes no alarm is caused, but spindle does not react.... Sometimes it does work. weird.

**Comment by Harvie** [comment]
Sometimes i get error:20
And i've found that M3/M4/M5 commands are relatively reliable when RPMs are not specified. But when i specify rpm eg. `M3 S10000` it causes issues.

**Comment by terjeio** [comment]
error:14 (or rather alarm:14? is from the plugin), error:9 (_not_ alarm:9) is likely from alarm:14 beeing active.

Wonder if disabling polling for the real-time report will make it more stable, can you try with commenting out this line?

https://github.com/terjeio/grblHAL/blob/61f09cd87cc5ec16fdf1711b90b76ea516ae3d15/plugins/spindle/huanyang.c#L178

It could be down to timing issues, I am not sure my code is 100% compliant with the ModBus spec (and perhaps the spindle is neither).

> Sometimes i get error:20

That suggests corrupted gcode. Noise?

Have you controlled the spindle sucessfully with grbl_esp32 before?

**Comment by Harvie** [comment]
> Wonder if disabling polling for the real-time report will make it more stable

Did not helped.

I also tried to comment out `system_raise_alarm(Alarm_Spindle);` which SEEMS to make things bit better, but i think it's just an illusion :-)

> That suggests corrupted gcode. Noise?

You might actually be true... I will have to do a lot of testing... I currently run my machine on 5V arduino and initialy it took me some time to fix the EMI issues. i guess the 3.3V ESP32 might be even more prone to such issues.

So if you don't have any other ideas. (eg. repeat each command 10 times hoping that at least one will be successful even during EMI. = might not be good for debugging, but also might make the overall system bit more robust)

I will try to hunt down the EMI problems... It will probably take some time...

**Comment by Harvie** [comment]
BTW what does `Pn:R` mean? I don't remember ever seeing this in classic GRBL. eg.:

`<Alarm|MPos:0.000,0.000,0.000|Bf:35,1022|FS:0,0|Pn:R>`

**Comment by terjeio** [comment]
Can you try with increasing the RX timeout, if the VFD is slow to respond then alarm 14 is raised:

https://github.com/terjeio/grblHAL/blob/61f09cd87cc5ec16fdf1711b90b76ea516ae3d15/drivers/ESP32/driver.c#L1574-L1577

> BTW what does Pn:R mean? I don't remember ever seeing this in classic GRBL.

It means the grbl reset pin is asserted - grblHAL defaults to normally closed switches for safety reasons (unlike grbl). Invert the input with `$14=1` to get rid of it (or short the reset pin to ground).

**Comment by phil-barrett** [comment]
> 3.3V ESP32 might be even more prone to such issues.

Definitely true.  especially if using NO inputs. (NC are a little better, btw)  That is why I am making 12V an option for the control inputs on my next board design. Much more resilient to EMI.

**Comment by Harvie** [comment]
Another thing i noticed:

0.) spindle running at eg. 2700 RPM
1.) i send for example S3000 and spindle does not do anything
2.) i send S3000 again and grbl responds with "ok", but does not re-send the S3000 to spindle, because it thinks that spindle is already at 3000 RPM even when it somehow missed the message in 1.)
3.) i can send S3000 how many times i want and grbl replies with OK, until i change to lets say S3200, then whole cycle repeats (based on whether or not the spindle got the message.)


> Can you try with increasing the RX timeout, if the VFD is slow to respond then alarm 14 is raised:

OK i will try that

**Comment by Harvie** [comment]
> Can you try with increasing the RX timeout

Tried rx_timeout = 400 and initialy it felt bit better. first 3 commands worked, but then the problems rised again... Also tried rx_timeout=100 and it feels same as with original 40...

**Comment by Harvie** [comment]
I tried shorter RS485 cable and baudrate of 4800 (lowest possible setting on VFD) and things seem even worse...

**Comment by Harvie** [comment]
Another thing i consider weird is that 1 or 2 commands after GRBL reset works and then something bad happens.
Maybe there are some partialy received messages in buffer screwing all subsequent communication?

**Comment by Harvie** [comment]
Just for reference, i am using following setup:

![image](https://user-images.githubusercontent.com/276504/99998547-23f20880-2dbf-11eb-8eed-7b108ccf6a79.png)

![image](https://user-images.githubusercontent.com/276504/99998409-e8efd500-2dbe-11eb-95e1-4df4f6530fec.png)

Except the IO pin numbers of esp32 are different (reflecting the *map.h file).

There might be several issues with this:

1.) i am not sure if MAX485 can use 3.3V (TTL both for Vcc and IO)
2.) i am not sure whether it's ok that i don't have any shielding. I just use ~50cm long twisted pair pulled out of UTP cable. Such unshielded cable was working for me when using dedicated USB-RS485 adapter, so i expect that to be OK.

**Comment by phil-barrett** [comment]
Max485 can run on 3.3V but you might need to change the differential resistor (Rt on the Max485 datasheet).

**Comment by Harvie** [comment]
> Max485 can run on 3.3V but you might need to change the differential resistor (Rt on the Max485 datasheet).

Thanks for the tip. I will certainly look into this!

**Comment by phil-barrett** [comment]
I should have been clearer. The MAX485 is spec'd to run on 5V.  There are people that run it on 3.3V. Somewhere (i.e. I do not remember where) I saw a discussion of 3.3V operation. I think (do not hold me to this) that Rt needed to be lower by the ratio of 3.3/5.

**Comment by Harvie** [comment]
> You might need to change the differential resistor (Rt on the Max485 datasheet)

Currently there is 120 ohm resistance when measured across the output pins of the module

> Also, both receiver and transmitter needed to be at the same voltage.

Ah. That might be the problem as i have no plans to modify the VFD to use 3.3V...
However i can run the MAX485 on 5V and use 1K to 10K resistor as protection for ESP32 rx input pin. Hopefully the 3.3V TX coming from ESP32 will be enough to drive input of MAX485 running at 5V.

**Comment by phil-barrett** [comment]
> However i can run the MAX485 on 5V and use 1K to 10K resistor as protection for ESP32 rx input pin. Hopefully the 3.3V TX coming from ESP32 will be enough to drive input of MAX485 running at 5V.

RX - I would use a 510/1000 divider. A 470/1000 divider will work at a slightly higher output voltage.
TX - the datasheet says Vih min is 2.0V so you should be good.

**Comment by terjeio** [comment]
> 2.) i send S3000 again and grbl responds with "ok", but does not re-send the S3000 to spindle, because it thinks that spindle is already at 3000 RPM even when it somehow missed the message in 1.)

It is coded that way, IMO better exception handling is required. There is two basic types of exception, timeout and exception reply from the VFD. Both result in alarm 14. I wonder what kind of exceptions you are getting.
To see that add a message report for the exception  code:

```
static void rx_exception (uint8_t code)
{
    report_message(uitoa(code), Message_Warning);
    system_raise_alarm(Alarm_Spindle);
}
```
in huanyang.c. 0 is timeout, I have to look up any other codes in the ModBus spec.

> Another thing i consider weird is that 1 or 2 commands after GRBL reset works and then something bad happens.
Maybe there are some partialy received messages in buffer screwing all subsequent communication?

If so the `stream->flush_rx_buffer()` call is not working. It is called before every message sent. Unlikely?

**Comment by Harvie** [comment]
> To see that add a message report for the exception code:

Ok, i will test that, but i am not near the machine now...

@phil-barrett btw what about shielding? is that required for RS485 cabling?

**Comment by phil-barrett** [comment]
> what about shielding? 

Can't hurt but twisted pairs and differential signaling should be fairly noise resistant to start with.

**Comment by terjeio** [comment]
Come to think of one possible reason behind the problems. You are using the direction pin, this has to be changed when all the data has left the UART shift register, not only the FIFO as I do now. I have to dig into the datasheet to see if I can find a way to be certain all bits are on the wire before switching the direction pin (or add a minimum delay after the transmit FIFO is empty). IIRC the iMXRT1062 sets a flag when transmission is complete, perhaps the ESP32 has one too.

**Comment by terjeio** [comment]
>...perhaps the ESP32 has one too.

There is a flag and an related interrupt too so for precise timing of the direction signal I am going to set it in the UART driver when a message is sent, enable the interrupt and reset it in the interrupt handler. It looks good on my scope, no randomness due to polling anymore. Will commit tomorrow.

**Comment by Harvie** [comment]
> You are using the direction pin

I beleive there is no way to use bidirectional RS485 communication without using direction pin.
You can get away without RX enable pin and supress loopback in ESP32 instead, which gives you oportunity to detect collisions. However you would still need properly timed direction signal for TX enable unless you are the only transmitter on the bus.

> I am going to set it in the UART driver when a message is sent, enable the interrupt and reset it in the interrupt handler.

Can't the ESP32 UART hardware be configured with direction pin number to handle this for you, so it does not have to be handled in software?

https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/uart.html#overview-of-rs485-specific-communication-options

**Comment by terjeio** [comment]
> I beleive there is no way to use bidirectional RS485 communication without using direction pin.

Well, it is working for me with the chinese tranceivers I bought from ebay - unmarked chips though... With auto direction detect. The Rx/Tx LEDs has been happily blinking for many hours now while talking to the simulator. No alarms raised.

> Can't the ESP32 UART hardware flow control be configured with direction (RTS) pin number to handle this for you, so it does not have to be handled in software?

It can, but requires a bit more digging into the datasheet? Pin assignment restrictions apply? Easy way first to see if this solves the issue.

Here are the changed files:

[GrblDriver.zip](https://github.com/terjeio/grblHAL/files/5589318/GrblDriver.zip)

**Comment by Harvie** [comment]
> With auto direction detect. 

With auto direction you have to use some kind of null byte supression/escaping in your protocol.

> It can, but requires a bit more digging into the datasheet? Pin assignment restrictions apply?

I don't really think this is the case as esp32 has means to internaly reroute the peripherals to different pins.
Actually the only thing i do is checking pin numbers against this table:
https://randomnerdtutorials.com/esp32-pinout-reference-gpios/

> Here are the changed files

Thanks, i will hopefully try that this evening...

**Comment by Harvie** [comment]
I've managed to hunt down the evidence of problem in the original code:

![image](https://user-images.githubusercontent.com/276504/100092741-dcb75680-2e56-11eb-84ec-3699c73bd872.png)

So i guess i can compare it with the new code...

**Comment by terjeio** [comment]
> I've managed to hunt down the evidence of problem in the original code:

Yep, this is from not waiting until the shift register is empty. I have updated the serial2txCount() function to add 1 to the count if transmit is not complete, this in addition to flipping the direction pin in the uart driver code.

**Comment by Harvie** [comment]
After update i am unable to find such issue in the waveform, i will try it with the real spindle later today:

![image](https://user-images.githubusercontent.com/276504/100094716-bcd56200-2e59-11eb-9db9-32a1ca02e50e.png)

**Comment by Harvie** [comment]
> Here are the changed files: GrblDriver.zip

Just tested on the machine and so far not a single issue

**Comment by Harvie** [comment]
There is some weird rounding going on... But that's rather minor issue. Eg.:

```
M3 S17000    -> 16999 on VFD LCD
M3 S17000.3 -> 16999 on VFD LCD
M3 S17000.4 -> 17000 on VFD LCD
M3 S17001    -> 17001 on VFD LCD
```

**Comment by terjeio** [comment]
Thanks for your effort!

> There is some weird rounding going on... But that's rather minor issue.

Nothing can be done about that?
```
        uint16_t data = (uint16_t)(rpm * 100 / 60); // send Hz * 10  (Ex:1500 RPM = 25Hz .... Send 2500)

        rpm_cmd.adu[1] = ModBus_WriteCoil;
        rpm_cmd.adu[2] = 0x02;
        rpm_cmd.adu[3] = data >> 8;
        rpm_cmd.adu[4] = data & 0xFF;
```
> With auto direction you have to use some kind of null byte supression/escaping in your protocol.

It does not seem to be required for the transceiver I've got. There are chips supporting auto direction control, e.g. [MAX13410E](https://datasheets.maximintegrated.com/en/ds/MAX13410E-MAX13413E.pdf), one would assume these do not require any protocol modifications. With auto direction the direction pin can be used for other purposes, important for the ESP32 since it has a limited number of pins available.

> Except for this, i would consider the code production ready...

IMO the code could be improved a bit. E.g. now I poll the VFD on every real time report (typ. every 200ms), if a valid reply is not received then what? Currently I raise alarm 14, is this how it should be? Should there be retries when status changes are requested and the spindle does not respond? A $-setting for baud rate?

For the ESP32 options for baud rate (if not implemented as a $-setting) and auto direction should be added to CMakeLists.txt.

**Comment by Harvie** [comment]
For ther record, i've been using cheapest esp32 devkit and cheapest max485 module with everything running at 3.3V, no shielding on RS485 data lines, just twisted wire (except for spindle 3-phase AC power wiring which absolutely has to be shielded). Didn't tested long cabling, but as far as ESP32 is in reasonable distance from VFD it should work. No 5V needed.

![image](https://user-images.githubusercontent.com/276504/100133176-b6111400-2e86-11eb-919a-01a37ef4b6e3.png)

**Comment by Harvie** [comment]
> if a valid reply is not received then what? Currently I raise alarm 14, is this how it should be?

It's not really a defined anywhere. I guess when spindle fails to start, we should probably not crash stalled endmill into the material, so it probably makes sense to stop the g-code execution in such case. But i would strongly suggest to make this runtime configurable, because sometimes i want to override the spindle VFD by setting it to manual instead of rs485 mode... for whatever reason... Or simply run the machine without VFD powered on.

I guess we can have some integer number to indicate number of retries before alarm, eg.:

0 - never fail
1 - fail immediately after 1 failed attempt
42 - fail after 42 failed attempts

>  A $-setting for baud rate?

Certainly good idea... The VFD itself has only 4 options: 4800, 9600, 19200, 34800 
But i guess being able to enter any arbitrary baudrate would be more future proof.

**Comment by Harvie** [comment]
> IMO the code could be improved a bit.

BTW it is questionable whether we should use the hardware flow control (RTS) to switch the transciever direction or handle the direction pin manualy. It takes some CPU cycles to do that manualy and might not have perfect timing. On the other hand if handled manualy it might be easier to port the code to other platforms (eg. STM32 or something with more GPIO pins).

> now I poll the VFD on every real time report (typ. every 200ms)

Would it make sense to make the polling interval configurable? Are there any possible benefits? Currently it is not completely clear to me why we even poll... You have to set the RPM and then poll the VFD to see if it got correctly configured, because it does not confirm the RPM directly when it get set?

>  uint16_t data = (uint16_t)(rpm * 100 / 60); // send Hz * 10  (Ex:1500 RPM = 25Hz .... Send 2500)

i might have a look into the VFD manual, whether it is possible to set RPM rather than frequency...

**Comment by terjeio** [comment]
> BTW it is questionable whether we should use the hardware flow control (RTS) to switch the transciever direction or handle the direction pin manualy. It takes some CPU cycles to do that manualy and might not have perfect timing. On the other hand if handled manualy it might be easier to port the code to other platforms (eg. STM32 or something with more GPIO pins).

I leave it up the driver implementation as this is handling the UART communication. The driver calls `modbus_init()` with a pointer to a structure holding pointers to the UART functions to use, one of these is `set_direction`, this is optional (could be NULL). If set then the plugin will call this function to set the direction pin as needed, e.g:
```
if(stream->set_direction)
    stream->set_direction(true);
```
So a driver implementer can choose how to handle this, either in the UART code or by providing a function to handle the pin manually. BTW porting the plugin code is not neccesary, it can be used for all processors if the driver has the required entry points (functions) implemented. 

> Would it make sense to make the polling interval configurable? Are there any possible benefits? 

Not that I can think of, and it would only complicate the code.  BTW it is kind of configurable already, by setting the sender polling rate for the real-time report.

> You have to set the RPM and then poll the VFD to see if it got correctly configured, because it does not confirm the RPM directly when it get set?

If the message setting the RPM is accepted one could assume the spindle is running at the requested speed... grblHAL has implemented functionality for "spindle at speed" to ensure this. This requires either an encoder connected to the spindle or some other way to get the actual RPM. IMO a good VFD should return the actual RPM when queried, not the programmed one - if so this can be used as input to the "spindle at speed" algorithm.

**Comment by Harvie** [comment]
> IMO a good VFD should return the actual RPM when queried, not the programmed one

Idealy it would be able to return both... the VFD has factory preset 20 second acceleration. That would mean you can only confirm that speed is correctly set after waiting for 20 seconds...

> This requires either an encoder connected to the spindle or some other way to get the actual RPM.

BTW. Have you considered plugin to do closed loop PID control of spindle? To drive spindle using RPM sensor (eg. 1 pulse per revolution) and PWM output (increasing PWM duty cycle when RPMs start dropping under load?). For me this would be preferred for smaller machines, where it would be overkill to use VFD. This is also very popular spindle setup in hobby CNC world:

1.) low voltage DC motor based spindles with PWM can be easily retrofited with RPM sensor
![image](https://user-images.githubusercontent.com/276504/100155947-68a59e80-2ea8-11eb-97f9-bc4efc7fe81a.png)

2.) cheap palm routers with brushed AC motors can be retrofited with RPM sensor and the PWM can be used to provide reference voltage for triac regulation:
http://www.vhipe.com/product-private/SuperPID-Home.htm

**Comment by terjeio** [comment]
> Idealy it would be able to return both... the VFD has factory preset 20 second acceleration. That would mean you can only confirm that speed is correctly set after waiting for 20 seconds...

20 seconds must be for some really heavy motors (or high RPM ones?), you should be able to reduce that quite a bit?

> BTW. Have you considered plugin to do closed loop PID control of spindle? 

There is code in the [MSP432 driver](https://github.com/terjeio/grblHAL/blob/master/drivers/MSP432/driver.c) for that. However, I have decided against a plugin for spindle sync and I guess the same reasons are valid against a closed loop spindle control plugin. So copy/pasting code is the way to add closed loop control to more drivers.  Some processor specific code is needed for getting the data for the RPM calculation too.

**Comment by Harvie** [comment]
> I have decided against a plugin for spindle sync and I guess the same reasons are valid against a closed loop spindle control plugin

Can i know why? For example the esp32 has independent pulse counting hardware peripheral, which would probably handle the RPM readout quite elegantly without interferring with interrupts and other stuff... (not to mention it has two cores, so one might be dedicated to the motion control to make sure it remains smooth)

**Comment by terjeio** [comment]
> Can i know why?

This is the main reason:

> For example the esp32 has independent pulse counting hardware peripheral...

and then:

> ... which would probably handle the RPM readout quite elegantly without interferring with interrupts and other stuff... 

which could be quite different from the interrupt based spindle sync implementations I have made. Spindle sync is a bit different though as this requires an index pulse in addition to the main encoder pulse train.

---


Each processor has different peripherals requiring HAL access if a plugins is to be made. For spindle sync and closed loop spindle RPM control this is perhaps half of the code needed. HAL overhead and lack of a RTOS are other reasons. So copy/paste of relatively few lines of code is the way to do it. To make it easier a template could be made.

What is readily available and handled by the core are settings for spindle pulses per revolution (PPR) and the PID values. These are enabled by setting a capability flag in the HAL structure. "Spindle at speed" likewise, a couple of code lines and setting a capability flag and the core takes care of the rest.

I am by no means negative to adding closed loop spindle control and would be happy if you want to give it a try, if it can result in a sensible plugin then I am all for that too.

**Comment by Harvie** [comment]
> if it can result in a sensible plugin then I am all for that too.

fair enough :)

Do you need any help with cleaning up the rs485 plugin? I have plans to write some better documentation for the esp32 build process, so it's easier on beginners. And perhaps i will try to look into the arduino ide support...

**Comment by terjeio** [comment]
> Do you need any help with cleaning up the rs485 plugin?

Too late, already done for the modbus plugin - settings added for baud rate (from a predefined list to reduce support due to misconfiguration) and RX timeout - and a hardcoded silent period after each message per ModBus spec (baud rate dependent). Switching the VFD on/off is a tad more complicated, have to think about that. A $-command?

> I have plans to write some better documentation for the esp32 build process, so it's easier on beginners. And perhaps i will try to look into the arduino ide support...

That would be nice.

**Comment by Harvie** [comment]
>        Nothing can be done about that?
>        uint16_t data = (uint16_t)(rpm * 100 / 60); // send Hz * 10  (Ex:1500 RPM = 25Hz .... Send 2500)

I've been thinking about it. For example these are calculations i've made in bc:

```
scale=1
((17000*100 / 60)) * (60/100)
16999.9
```

We can decode the frequency back to RPM (using the same algorithm as used in VFD) and if the floor() of the resulting value is not equal to our original RPM, the simply increment the frequency slightly.

This is very naive approach, but proves that this can be figured out.

This can probably be solved using some precision juggling magic. My gut feels like adding 0.5 to target RPM before calculation might do the trick... but i am not sure yet.

**Comment by terjeio** [comment]
`        uint32_t data = lroundf(rpm * 100.0f / 60.0f); // send Hz * 10  (Ex:1500 RPM = 25Hz .... Send 2500)`

is better, casting did not round - it behaves like floor. I was not aware of that.

> We can decode the frequency back to RPM (using the same algorithm as used in VFD) and if the floor() of the resulting value is not equal to our original RPM, the simply increment the frequency slightly.

grblHAL adds the actual RPM as a third parameter to the `FS:` real time report element when a function for `hal.spindle.get_data` is provided by the driver or a plugin. The Huanyang plugin did provide any so I had to correct that. grblHAL reports the spindle RPM back as an int casted from float, I'll change that to use `lroundf` instead in the next build.

> BTW does M3 Sxxx realy have to accept decimal numbers or it's just non-standart grbl extension?

Good question. NIST has this clue on page 94:

> double GET_EXTERNAL_SPEED();
Return the system value for the spindle speed setting in revolutions per minute (rpm). The
actual spindle speed may differ from this.

I have decided to use int for reporting in grblHAL and as the input field format in ioSender. float is still used internally in the core but not necessarily by all drivers and plugins. Reasoning behind that: 8-bit grbl uses PWM for RPM control, with an 8-bit timer for PWM too boot. Most (if not all) grblHAL drivers has at least 16 bit PWM. In both cases commanding exact RPMs with fractional values has no mapping to the real world, even with closed loop spindle control. A Huanyang VFD is similar as there are max. 2^16 discrete RPMs available.

How many grbl users have (or will get) a spindle that is possible to control precisely? And has a need for that?

**Comment by Harvie** [comment]
> is better, casting did not round - it behaves like floor. I was not aware of that.

Well it depends on how bad the VFD firmware is. Maybe they have some issues with precision as well so that would mean we have to add opposite imprecission in our code. In the past i had horror experiences with chinese firmware, including precision problems (different devices, not HY VFDs), so nothing can really suprise me...

Anyway i will thest the suggested code and tell you what the results are...

> How many grbl users have (or will get) a spindle that is possible to control precisely? And has a need for that?

I highly doubt that deviation of 1 RPM will make any difference unless you are milling at speed of 5 RPMs :-) To be honest only reason why i am looking into this small deviation is my OCD, i practice i don't really care if i cut at 8000RPM or 7999RPM, but having both displays show the same value will certainly make this look more professional and trustworthy...

**Comment by Harvie** [comment]
This still gives me wrong RPM

```C
#include <math.h>
uint16_t data = (uint16_t)(rpm * 100 / 60); // send Hz * 10  (Ex:1500 RPM = 25Hz .... Send 2500)
```

This seems to fix the issue:

```C
uint16_t data = (uint16_t)((rpm+0.5) * 100 / 60); // send Hz * 10  (Ex:1500 RPM = 25Hz .... Send 2500)
```

**Comment by terjeio** [comment]
New commit made to the test branch. Settings for modbus baud rate and rx timeout added, use `$help modbus` for details.

The commit uses lroundf to fix the RPM issue, and requires lroundf in report.c to get back to original value. If this does not work as it should I'll take a look at it again with your proposal above.

Missing report of the actual RPM read from the spindle added to the real time report. Other issues fixed to, the main was hang on a soft reset. Enumeration data for the new settings added (to `$ES` and `$EG`), can be used by senders for the settings UI.

> This still gives me wrong RPM

Where? In your OCD?

**Comment by Harvie** [comment]
> Where? In your OCD?

Well. I just issue `M3 S8000`, the VFD starts the spindle and after second or two it settles at 7999 RPMs. (VFD has it's own display showing the RPMs)

**Comment by Harvie** [comment]
> Switching the VFD on/off is a tad more complicated, have to think about that.

Idealy i'd like to be able to use modbus to control spindle and PWM output to control laser, so i can have both on single machine...

**Comment by terjeio** [comment]
> Yes. I just issue M3 S8000, the VFD starts the spindle and after second or two it accelerates to 7999 RPMs.

Should be 7998, with your rounding it should be 8004. So there is some magic going on - imprecise PID or faked RPM in the VFD display? Wonder what the measured RPM is? The error is less with lroundf...

BTW PD003 can set the frequency in steps of 0.01Hz, commanded via ModBus only 1Hz.

> I can pull the plug of VFD so there's no need to actualy turn it off. 

A $-command or a setting (use the $32 setting for laser mode?) to switch mode? Reset of controller required? The VFD code "inserts" itself by claiming a number of HAL entry points, these _can_ be restored back to the default PWM mode of operation.  However, I do not like the idea of an unresponsive VDF triggering that automatically... Checking for laser mode on boot could be easy to implement, IMO not problematic as nobody would want to use a VFD to control a laser?

**Comment by Harvie** [comment]
> So there is some magic going on - imprecise PID or faked RPM in the VFD display?

I don't know, but the code i've provided works well. To be honest... As long as it does work properly, i don't even care how "chinese rounding" in VFD works to save myself from that horror. They probably just truncate the resulting RPM in VFD instead of rounding, so the offset of 0.5 fixes that.

> A $-command or a setting (use the $32 setting for laser mode?) to switch mode?

Yes, i guess the $32 might be a good fit.

Also i've been thinking about some magic tool number, so we can switch between spindle and laser using some magic number, eg.: M6 T9999 to switch to the laser and any other tool number to switch back...

That way i would be able to laser-engrave and cut the piece in one g-code file...

But the $32 is stored across grblhal reboots, so it makes sense as well...

> Reset of controller required?

I don't think that's necessary...

> IMO not problematic as nobody would want to use a VFD to control a laser?

I am not aware of any laser on hobby CNC market that can be driven by VFD :-)

**Comment by terjeio** [comment]
> Also i've been thinking about some magic tool number, so we can switch between spindle and laser using eg.: M6 T9999 to switch to the laser and any other tool number to switch back...

Use $32 to switch - the plugin already subscribes to the settings changed event so that would be a "clean" way to do it. Tool change logic that is not standard is not something that I will add to grblHAL. You can of course write your own custom plugin though.

For the RPM issue I'll stick to lroundf as this fits better with the rounding in report.c - I am not going to change report.c to fit the way Huanyang do the rounding. There may be other VFDs behaving differently, perhaps even their v2 protocol does?

**Comment by Harvie** [comment]
Even LinuxCNC has some weird compensation for this in their code:

https://github.com/bebro/linuxcnc-huanyang-vfd/blob/emc2/hy_vfd.c#L205

**Comment by terjeio** [comment]
I think I have a solution to dual spindle. It will not work without additional code and/or a board file since pins has to be assigned for both spindle types. If reassigning pins dynamically then external hardware is required? If the driver/board map is capable of handling two spindles then `hal.driver_cap.dual_spindle` has to be set before huanyang_init() is called.

Note that there could be issues surfacing around switching spindle, likely so if switching is done while the active one is running.
To be continued - I'll take a short break from coding.

**Comment by Harvie** [comment]
bdring grbl port also has different approach to frequency calculation:
https://github.com/bdring/Grbl_Esp32/pull/544/files#diff-7fed6b73a3c9ce8053c389e6c1c6eabbc507cce0fede33d2ff6269b3b25ce15eR68

**Comment by terjeio** [comment]
> bdring grbl port also has different approach to VFD frequency calculation:

Same as the one grblHAL had originally:

`uint16_t value = (uint16_t)(rpm * 100 / 60);`

The v2 (H2A) protocol is different, this includes the max rpm setting read from the VFD:

`uint16_t data = (uint32_t)(rpm) * 10000UL / rpm_max;`

Seems like LinuxCNC does the same for the v1 protocol. I prefer to follow manufacturer specifications, but there are none to be found? I have some documents but they are useless. BTW my latest version for v1 gives the least error, displays correct in grblHAL and wrong in the VFD display. Huanyang should fix their firmware? ;-)

**Comment by Harvie** [comment]
Btw the `lroundf()` and `rpm+.5` compensation can be used both at the same time...

**Comment by terjeio** [comment]
> Why does VFD frequency value leave the huanyang subsystem?

"Spindle at speed" handling, display of actual RPM in the sender. Don't know what the Huanyang reports back (programmed or actual - hopefully actual), will have to check the manual for clues.

**Comment by Harvie** [comment]
@terjeio interresting summary:
https://github.com/bdring/Grbl_Esp32/pull/544#issuecomment-734745004

**Comment by atlaste** [comment]
The manuals are a PITA. I think all that we found are still referenced in our VFD Huanyang code; we ended up using all of them... Fact is that most of them are incomplete and registers don't always report what they should do. It might even be possible that actual frequency is there somewhere in the Huanyang VFD, but we just haven't found it. I don't own a v1 huanyang, otherwise I would consider setting it at some speed, and then attempting to query all the registers to see if one has the right information. 

Details:
- Huanyang: https://github.com/bdring/Grbl_Esp32/blob/Devt/Grbl_Esp32/src/Spindles/HuanyangSpindle.cpp line 42-...
- H2A: https://github.com/bdring/Grbl_Esp32/blob/Devt/Grbl_Esp32/src/Spindles/H2ASpindle.md

In any event, if you would happen to find the actual frequency, please let us know!

I wouldn't use floating point to be honest. I've seen VFD spindles with very high RPM values (200K+), so rounding issues will probably occur.

**Comment by scottbez1** [comment]
Jumping in late here, as I've just been digging into the Huanyang spindle driver in Grbl_Esp32 while building up my machine.

My interpretation of the Huanyang manual is that the "correct" conversion between RPM and Frequency shouldn't hard-code any constants and should probably be based on PD values from the VFD, though which one(s) is somewhat unclear.

The internet tells me the basic RPM vs Frequency formula is `RPM = (120 * Frequency) / # of poles in the motor` - in which case you could read PD143 (Motor Pole Number) to convert between the two -- `RPM = 120 * Freq / PD143`

Separately, I see PD144 (Rated Motor Revolution), which sounds like what they might use to calculate the RPM that's displayed on the VFD? So an alternative approach seems like `RPM = Freq * PD144 / 50` (the 50 is because PD144 is defined as the RPM at 50Hz)
![Screenshot from 2020-12-09 03-26-41](https://user-images.githubusercontent.com/414890/101624508-dd132e00-39ce-11eb-856f-25f36f3fad6d.png)

In practice, I suspect most of the hobby usages of the Huanyang VFD on CNC machines are for the same common 2-pole spindles, so the hardcoded constants probably work fine for the vast majority of hobby CNC machines with Huanyang VFDs...

Separately, I see the issue of actual RPM (as opposed to the _set_ RPM or frequency) discussed above. This can be found by sending a function 4 (Read status value) message, requesting status field 3 (RoTT). In the version of the manual I was looking at, the example is not correct... I've noted the discrepancy [on my work-in-progress Huanyang protocol document](https://paper.dropbox.com/doc/Huanyang-VFD-modbus-protocol--BA_2NvOtb~6wRqJd4rzshRLwAg-836X8EruAxH21PFe0Z2Um#:uid=456566706052355512021141&h2=04:-Read-status-value) and I've confirmed that reading this value works on my Huanyang VFD.

**Comment by terjeio** [comment]
@scottbez1 Thanks for you input. I guess I need to buy a Huanyang VFD to be able to dig into this myself...
[LinuxCNC ](https://github.com/bebro/linuxcnc-huanyang-vfd/blob/emc2/hy_vfd.c#L205) reads a bunch of settings on startup that I assume is used later on. Perhaps some useful info can be extracted from that implementation too?

**Comment by Harvie** [comment]
> I guess I need to buy a Huanyang VFD to be able to dig into this myself...

What spindle do you currently run on your machine?

**Comment by terjeio** [comment]
> What spindle do you currently run on your machine?

Kress FM1050 for the router (has Mach3/ethernet Smootstepper - not going to change to grbl).
Small 48V DC motor for the mini router.
Washing machine motor with no brandname VFD for the lathe (0-10V control), no RS-485 interface AFAIKT.

**Comment by Harvie** [comment]
I've just checked test branch and i see you've added following settings:

```
$374: ModBus baud rate:
0 - 2400
1 - 4800
2 - 9600
3 - 19200
4 - 38400
5 - 115200
$375: ModBus RX timeout in milliseconds, range: 50 - 250
```

can you please add option to set `$375=0` in order to ignore unresponsive VFD? I have several use cases:
 * if you are optimistic you can save the RX pin for some other use and hope the modbus messages sent on TX pin will get received.
 * i usualy flash and test the grblHAL at home and when everything seems OK i bring it to the workshop to install it into the machine. So i get ALARM:14 because i don't have VFD at home, it would be super useful for me to be able to test everything but ignore missing VFD.
 * For some reason i might want to be able to configure VFD frequency manualy by the buttons on device, therefore i need to be able to tell grblHAL to ignore it.

**Comment by raenji-sk** [comment]
Bump.

**Comment by Harvie** [comment]
I don't think this will work without modification with other brands, but the boilerplate code is there and you can probably modify it if you know what modbus commands your unit expects.

**Comment by raenji-sk** [comment]
Ok, thanks. I'll try to look into it when or if I decide to upgrade the spindle. Or bother you here for help 😅.

**Comment by phil-barrett** [comment]
It needs an RS485 adaptor.  That is something I am planning to do.  That is one of the reasons the new Pro board has mounting holes for a UART daughter card.

**Comment by terjeio** [comment]
> But I am in no rush so there's time to learn something new perhaps.

I think we can sort this out together, for basic control not much code needs to be changed - the largest part of the job is to figure out which commands to send and testing.

**Comment by phil-barrett** [comment]
Here are 2 cards that might work for you.  I have the first (ANMBEST) but think the chip is not really a max485
https://www.amazon.com/ANMBEST-Transceiver-Arduino-Raspberry-Industrial-Control/dp/B088Q8TD4V/
https://www.amazon.com/HiLetgo-Reciprocal-Hardware-Automatic-Converter/dp/B082Y19KV9/

**Comment by MisterDiz** [comment]
I'm attempting to get my Huanyang VFD working with grblHAL on a Teensy 4.1

Is this implemented for the Teensy? 

I have it defined in _my_machine.h_. Is it then just a case of defining the pins in the _T41U5XBB_map.h_?
Do I comment out the existing spindle definitions with something like...

#if SPINDLE_HUANYANG 
#define UART2_RX_PIN            (12u)
#define UART2_TX_PIN            (13u)
#define MODBUS_DIRECTION_PIN    (11u)
#define MODBUS_BAUD             19200
#endif

I also saw this in the driver.h file for the ESP32, is it necessary?

#if SPINDLE_HUANYANG
#define RS485_DIR_ENABLE 1
#else
#define RS485_DIR_ENABLE 0
#endif

Am I then good to go after plugging in my 3.3v TTL to RS485 adapter?

**Comment by dresco** [comment]
> I'm attempting to get my Huanyang VFD working with grblHAL on a Teensy 4.1
> Is this implemented for the Teensy?

Yes, but is easier than that. You will just need to define VFD_ENABLE=1 and SPINDLE_RPM_CONTROLLED in your my_machine.h (or in platformio.ini if building from that)

Note that you will want to pull the latest code from github.com/grblHAL/, as this repo is the old location and out of date..

**Comment by MisterDiz** [comment]
Ha! Nothing in my world is ever easy ;P

Thanks for the quick reply Dresco, I'm now with a less archaic version, changed those 2 lines and compiled successfully :)

Should I use the "SPINDLE_ENABLE_PIN" for RX
and "SPINDLE_PWM_PIN" for TX?

Presumably there's nothing else I need to change?

**Comment by dresco** [comment]
It will default to Serial1, which I think is pin 0 for RX and pin 1 for TX. (The RS485 adapters I'm using don't have an enable line, so not sure what pin that would be mapped to if you need it).

Are you using Phils breakout board? If so, just hook your 3.3v TTL to RS485 adapter up to the serial header & should be good to go..

Cheers.

**Comment by MisterDiz** [comment]
Cool, will give that a go. I'm using a slightly ghetto looking DIY BOB, currently I'm having post upload 'GRBL not found'/'offline' issues with the new GrblHAL when connecting to my all my G-Code senders. :/

Thanks for your help

**Comment by MisterDiz** [comment]
Ok, so I have the latest GrblHAL up and running and active without spindle, when attempting to enable RS485 control, my sender can see the Teensy but I cannot get communication.

I have the spindle plug in folder with the 3 files and the only code changes I've made are by adding 
#define VFD_ENABLE          1 // Set to 1 or 2 for Huanyang VFD spindle. More here https://github.com/grblHAL/Plugins_spindle
#define SPINDLE_RPM_CONTROLLED 
to _my_machine.h_

Apologies for the school boy errors, what am I missing?

**Comment by phil-barrett** [comment]
Have you gone through the [first run settings](https://github.com/terjeio/grblHAL/wiki/First-Run-Grbl-Settings)? It is fairly common for a different configuration to return previous settings to default.  Try grounding the halt pin - the default is NC.

**Comment by MisterDiz** [comment]
Thanks Phil, yes, I inverted all the pins for $5, $6 and $14 on my first setup, the machine started with no alarms and worked as it should - my problem is that when I define VFD_ENABLE 1 and SPINDLE_RPM_CONTROLLED in _my_machine.h_, after reboot I get an offline status. Commenting these two lines out gives me my machine back but obviously no VFD control.

Do I need to comment out the default spindle control settings or are they ignored? (I was fearing this would lead to compile issues)

**Comment by terjeio** [comment]
> Ok, so I have the latest GrblHAL up and running and active without spindle, when attempting to enable RS485 control, my sender can see the Teensy but I cannot get communication.

Is this the version form the [new repository](https://github.com/grblHAL/iMXRT1062)? I am about to archive this to avoid confusion...

>  after reboot I get an offline status.

Are you connecting to the controller via UART or USB? "Oflline status" is from a sender, does the controller respond if you connect with a terminal application such as puTTY?

**Comment by MisterDiz** [comment]
Hi Terjeio, yes, after the advice of Dresco, I am now on the correct latest build from the other repository :)

My Teensy 4.1 is connected by USB, senders can identify a port and open a connection, but Grbl communication isn't ready. I've not tried to ssh, what address would I use?

**Comment by terjeio** [comment]
> I've not tried to ssh, what address would I use?

Connect via the USB com port. Do you get a response when sending `?` (for getting a real-time report)?

If the the VFD is not set up correctly or the baud rate is wrong the controller will start up with alarm 14 active. Baud rate is set with `$374`. Use `$help modbus` from the terminal to get info about settings if you can communicate with the controller.

**Comment by MisterDiz** [comment]
Good news, whilst PuTTY failed to connect (no status or position was returned), the problem was that I was adding SPINDLE_RPM_CONTROLLED to _my_machine.h_ rather than uncommenting in _grbl/config.h_

Controller now boots and connects :)

Thanks also for the $modbus tip, I am fiddling with the hardware before connecting it all up. 

In my board map, are the Spindle enable, direction and PWM pins redundant? (so I should uncomment them?)
Conversely, if I leave the enable pin defined, will it be available to control other relays?

**Comment by terjeio** [comment]
> In my board map, are the Spindle enable, direction and PWM pins redundant? (so I should uncomment them?)
Conversely, if I leave the enable pin defined, will it be available to control other relays?

Yes. However, if the PWM pin definition is missing it will generate a compiler error unless you update to the version just committed.

**Comment by dresco** [comment]
> the problem was that I was adding SPINDLE_RPM_CONTROLLED to _my_machine.h_ rather than uncommenting in _grbl/config.h_

Apologies, that probably came from my suggestion didn't it. I pass the defines in from IDE options, so had overlooked where it's set..

**Comment by MisterDiz** [comment]
No need to apologise Dresco. All is good. I now have full spindle talkie talkie and no release of magic smoke! 👍 

Thanks again for all the help everyone.
